build/build.c

```diff
@@ -196,9 +198,109 @@ rpmRC doScript(rpmSpec spec, rpmBuildFlags what, const char *name,
     return rc;
 }
 
+static rpmRC doBuildRequires(rpmSpec spec, int test) {
+    char * buildDir = rpmGenPath(spec->rootDir, "%{_builddir}", "");
+    char *scriptName = NULL;
+    char * buildCmd = NULL;
+    FD_t fd = NULL;
+    FILE * fp = NULL;
+    StringBuf sb_stdout = NULL;
+    int argc = 0;
+    const char **argv = NULL;
+
+    int outc;
+    ARGV_t output = NULL;
+
+    rpmRC rc = 1; /* assume failure */
+
+    if (!spec->buildrequires) {
+	rc = RPMRC_OK;
+	goto exit;
+    }
+
+    /* write script to disk */
+    fd = rpmMkTempFile(spec->rootDir, &scriptName);
+    if (Ferror(fd)) {
+	rpmlog(RPMLOG_ERR, _("Unable to open temp file: %s\n"), Fstrerror(fd));
+	goto exit;
+    }
+
+    if ((fp = fdopen(Fileno(fd), "w")) == NULL) {
+	rpmlog(RPMLOG_ERR, _("Unable to open stream: %s\n"), strerror(errno));
+	goto exit;
+    }
+
+    fprintf(fp, "cd '%s'\n", buildDir);
+
+    if (spec->buildSubdir)
+    	fprintf(fp, "cd '%s'\n", spec->buildSubdir);
+    fprintf(fp, "%s", getStringBuf(spec->buildrequires));
+    (void) fclose(fp);
+
+    if (test) {
+	rc = RPMRC_OK;
+	goto exit;
+    }
+
+    /* execute script */
+    buildCmd = rpmExpand("/bin/sh", " ", scriptName, NULL);
+    (void) poptParseArgvString(buildCmd, &argc, &argv);
+    rpmlog(RPMLOG_NOTICE, _("Executing(buildreqs): %s\n"), buildCmd);
+    if ((rc = rpmfcExec((ARGV_const_t)argv, NULL, &sb_stdout, 1, spec->buildSubdir)))
+	goto exit;
+
+    /* add results to requires of the srpm */
+    argvSplit(&output, getStringBuf(sb_stdout), "\n\r");
+    outc = argvCount(output);
+
+    if (!outc) {
+	rc = RPMRC_OK;
+	goto exit;
+    }
+
+    for (int i = 0; i < outc; i++) {
+	parseRCPOT(spec, spec->sourcePackage, output[i], RPMTAG_REQUIRENAME, 0, 0, addReqProvPkg , NULL);
+    }
+
+    rpmdsPutToHeader(*packageDependencies(spec->sourcePackage, RPMTAG_REQUIRENAME), spec->sourcePackage->header);
+
+    if (rc == 0) {
+	parseRCPOT(spec, spec->sourcePackage, "DynamicBuildRequires = 4.15", RPMTAG_PROVIDENAME, 0, 0, addReqProvPkg , NULL);
+    }
+
+ exit:
+    free(argv);
+    free(buildCmd);
+    free(output);
+    free(buildDir);
+    return rc;
+}
+
+static rpmRC doCheckBuildRequires(rpmSpec spec, int test) {
+    /* recheck the dependencies */
+    rpmRC rc = RPMRC_OK;
+    rpmts ts = rpmtsCreate();
+    (void) rpmtsSetRootDir(ts, spec->rootDir);
+    rpmtsSetFlags(ts, rpmtsFlags(ts) | RPMTRANS_FLAG_NOPLUGINS);
+
+    rpmps ps = rpmSpecCheckDeps(ts, spec);
+
+    if (ps) {
+	rpmlog(RPMLOG_ERR, _("Failed build dependencies:\n"));
+	rpmpsPrint(NULL, ps);
+    }
+    if (ps != NULL)
+	rc = RPMRC_MISSINGBUILDREQUIRES;
+    rpmpsFree(ps);
+    rpmtsFree(ts);
+
+    return rc;
+}
+
```

Comment:
The transaction set creation etc looks out of place in the build-code, and this duplicates similar check in rpmbuild.c
Moreover, build-require checking is now rather inconsistent as one part is done here as a regular part of the build process but the first half is a completely separate logic in rpmbuild.c.

One possibility at unifying these would be moving the first part of builddep check from rpmbuild.c to take place through rpmSpecBuild() too. The transaction needs to be passed in by the caller somehow, perhaps through rpmBuildArguments_s.

Reply:
Hmm, actually if the buildrequire-checking is moved entirely to librpmbuild-side then it's probably better to make the transaction set a direct argument of rpmSpecBuild() to make the change explicit.